add 00_ingest and 01_cases tasks

neverett opened the pull request:
WIP for now because i'd like to add type hints and a bit more inline documentation, and update READMEs, but feel free to comment!

In this PR:

- a new 00_ingests task with a makefile that matches the one in 02_allegations, and scripts that correspond to the steps in the makefile. this task copies rows from the case CSVs in `data/case_files/` to the `cases_raw` table.
- a new 01_cases task, also with an appropriate makefile and scripts. this task deduplicates rows in the `cases_raw` table, cleans the data in the intermediate `cases_raw_deduplicated` table, and inserts the cleaned data into the `cases` table.
- a refactored db_cnx() script in `sql.py` that defaults to the connection settings in `db_config.py` and the default `pymysql.cursors.Cursor` cursor; those defaults can be overridden anywhere `db_cnx()` is called.

additionally, i refactored the scripts in my tasks that connect to the DB to use a context manager to handle DB connections so we don't have to explicitly close connections or cursors.

tgj505 commented on .gitignore:
is `db_connect.py` meant to be in here twice?

neverett replied:
oops, no, good catch!

tgj505 commented on sql/dedup_cases_raw.sql:
```diff
@@ -0,0 +1,10 @@
+USE `nlrb_data`;
+
+DROP TABLE IF EXISTS `cases_raw_deduped`;
```
instead of the `DROP TABLE IF EXISTS...; CREATE TABLE` format, the `cases_raw.sql` and `cases.sql` use `CREATE TABLE IF NOT EXISTS`. Does it matter?

neverett replied:
in this case, i think we do want to drop this table before proceeding -- if there's an existing `cases_raw_deduped` table when this SQL script runs, it will (probably) contain deduplicated rows from the previous run, which could be out of date. (and since we'll likely only be running the 01_cases task when the `cases_raw` table changes, it's very likely any existing `cases_raw_deduped` table would be stale.)

tgj505 commented on Common/app_config.py:
would it be more generalizable to use some os pathing for these common directories? e.g., in the `common.py` files there's something like:
`os.path.join(os.path.dirname(os.path.abspath(__file__)), '../..')`

neverett commented on tasks/00_ingest/post.py:
```diff
@@ -0,0 +1,33 @@
+                    print(f'Expected {app_config.cases_raw}',
+                          'table to be populated,',
+                          'found 0 records')
+            except pymysql.err.ProgrammingError as e:
```
it could be useful to catch multiple exception types here, something like:
```
except (pymysql.err.ProgrammingError, pymysql.err.OperationalError) as e:
    ...etc...
```

i think that would probably cover the [error types](https://github.com/PyMySQL/PyMySQL/blob/main/pymysql/err.py) we're most likely to encounter here (and could be incorporated into other db-connection heavy scripts) -- what do you think?

neverett replied:
(in general, we should probably rethink exception handling in the scripts that do a lot of DB operations, but it might make sense to do that in the set of changes needed to switch to postgres, since i think the psycopg2 error classes are slightly different.)

neverett commented:
major changes in my most recent set of updates to this branch:
* task 00_ingest:
  * added more inline documentation and type hints, refactored Makefile
* task 01_cases:
  * switched from SQL to polars to deduplicate case data, which required a slew of supporting changes, namely:
    * updated `cases.sql` and `cases_raw.sql` to switch back to `varchar` from `text` to avoid annoying processing errors with polars
    * updated requirements.txt with new requirements (**note:** a big change here is that you need pandas >= 2.0.0, since there's a bug in earlier versions of pandas that makes writing a polars dataframe to the DB fail.)

tgj505 commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
 venv/
+__pycache__/
+.DS_Store
+db_config.py
```
i think line 5 should be `Common/db_config.py` ?

neverett replied:
i think git just matches on the patterns in the .gitignore file, so all db_config.py files will be ignored. i've since removed the duplicate db_config.py entry; i think that was a copy/paste error on my part.
